File: src-tauri/macos/native/LoginItem.m
#import <ApplicationServices/ApplicationServices.h>
#import <Foundation/Foundation.h>
#import <ServiceManagement/ServiceManagement.h>
#import <libproc.h>
#import <stdlib.h>
#import <string.h>
#import <unistd.h>

typedef NS_ENUM(int, ChiriLoginItemStatus) {
  ChiriLoginItemStatusUnsupported = -1,
  ChiriLoginItemStatusNotRegistered = 0,
  ChiriLoginItemStatusEnabled = 1,
  ChiriLoginItemStatusRequiresApproval = 2,
  ChiriLoginItemStatusNotFound = 3,
};

static char *ChiriCopyCString(NSString *value) {
  const char *utf8 = value.UTF8String;
  return utf8 == NULL ? NULL : strdup(utf8);
}

static void ChiriSetError(char **output, NSString *message) {
  if (output == NULL) {
    return;
  }

  *output = ChiriCopyCString(message);
}

static BOOL ChiriIsOperatingSystemAtLeast(NSInteger majorVersion, NSInteger minorVersion) {
  NSOperatingSystemVersion version = {
      .majorVersion = majorVersion,
      .minorVersion = minorVersion,
      .patchVersion = 0,
  };
  return [[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:version];
}

static int ChiriLoginItemStatusFromServiceStatus(SMAppServiceStatus status) {
  switch (status) {
  case SMAppServiceStatusEnabled:
    return ChiriLoginItemStatusEnabled;
  case SMAppServiceStatusRequiresApproval:
    return ChiriLoginItemStatusRequiresApproval;
  case SMAppServiceStatusNotFound:
    return ChiriLoginItemStatusNotFound;
  case SMAppServiceStatusNotRegistered:
  default:
    return ChiriLoginItemStatusNotRegistered;
  }
}

static BOOL ChiriParentProcessPathContainsLoginWindow(void) {
  char parentPath[PROC_PIDPATHINFO_MAXSIZE] = {0};
  int pathLength = proc_pidpath(getppid(), parentPath, sizeof(parentPath));
  if (pathLength <= 0) {
    return NO;
  }

  return strstr(parentPath, "/loginwindow.app/") != NULL;
}

void chiri_macos_login_item_free_string(char *value) {
  free(value);
}

int chiri_macos_was_launched_as_login_item(void) {
  NSAppleEventDescriptor *event =
      [NSAppleEventManager sharedAppleEventManager].currentAppleEvent;
  if (event != nil && event.eventID == kAEOpenApplication) {
    NSAppleEventDescriptor *launchDescriptor =
        [event paramDescriptorForKeyword:keyAEPropData];
    if (launchDescriptor != nil &&
        launchDescriptor.enumCodeValue == keyAELaunchedAsLogInItem) {
      return 1;
    }
  }

  return ChiriParentProcessPathContainsLoginWindow() ? 1 : 0;
}

int chiri_macos_login_item_status(void) {
  if (ChiriIsOperatingSystemAtLeast(13, 0)) {
    return ChiriLoginItemStatusFromServiceStatus([SMAppService mainAppService].status);
  }

  return ChiriLoginItemStatusUnsupported;
}

int chiri_macos_login_item_enable(char **outputError) {
  if (ChiriIsOperatingSystemAtLeast(13, 0)) {
    int status = chiri_macos_login_item_status();
    if (status == ChiriLoginItemStatusEnabled ||
        status == ChiriLoginItemStatusRequiresApproval) {
      return 1;
    }

    NSError *__autoreleasing error = nil;
    if ([[SMAppService mainAppService] registerAndReturnError:&error]) {
      return 1;
    }

    ChiriSetError(outputError, error.localizedDescription ?: @"Unknown login item error");
    return 0;
  }

  ChiriSetError(outputError, @"Launch at login requires macOS 13 or later.");
  return 0;
}

int chiri_macos_login_item_disable(char **outputError) {
  if (ChiriIsOperatingSystemAtLeast(13, 0)) {
    int status = chiri_macos_login_item_status();
    if (status == ChiriLoginItemStatusNotRegistered ||
        status == ChiriLoginItemStatusNotFound) {
      return 1;
    }

    NSError *__autoreleasing error = nil;
    if ([[SMAppService mainAppService] unregisterAndReturnError:&error]) {
      return 1;
    }

    ChiriSetError(outputError, error.localizedDescription ?: @"Unknown login item error");
    return 0;
  }

  ChiriSetError(outputError, @"Launch at login requires macOS 13 or later.");
  return 0;
}

File: src-tauri/swift/Sources/ChiriMacOSBridge/LoginItem.swift
import ApplicationServices
import Darwin
import Foundation
import ServiceManagement

private enum LoginItemStatus: Int32 {
    case unsupported = -1
    case notRegistered = 0
    case enabled = 1
    case requiresApproval = 2
    case notFound = 3
}

private let procPidPathInfoMaxSize = 4096

private func copiedCString(_ value: String) -> UnsafeMutablePointer<CChar>? {
    strdup(value)
}

private func setError(
    _ output: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    _ message: String
) {
    output?.pointee = copiedCString(message)
}

@available(macOS 13.0, *)
private func mainAppServiceStatus() -> LoginItemStatus {
    switch SMAppService.mainApp.status {
    case .enabled:
        return .enabled
    case .requiresApproval:
        return .requiresApproval
    case .notFound:
        return .notFound
    case .notRegistered:
        return .notRegistered
    @unknown default:
        return .notRegistered
    }
}

@_cdecl("chiri_macos_login_item_free_string")
public func chiriMacOSLoginItemFreeString(_ value: UnsafeMutablePointer<CChar>?) {
    if let value {
        free(value)
    }
}

@_cdecl("chiri_macos_was_launched_as_login_item")
public func chiriMacOSWasLaunchedAsLoginItem() -> Int32 {
    if let event = NSAppleEventManager.shared().currentAppleEvent,
       event.eventID == kAEOpenApplication,
       let launchDescriptor = event.paramDescriptor(forKeyword: keyAEPropData),
       launchDescriptor.enumCodeValue == keyAELaunchedAsLogInItem
    {
        return 1
    }

    return parentProcessPathContainsLoginWindow() ? 1 : 0
}

private func parentProcessPathContainsLoginWindow() -> Bool {
    var parentPath = [CChar](repeating: 0, count: procPidPathInfoMaxSize)
    let pathLength = parentPath.withUnsafeMutableBufferPointer { buffer -> Int32 in
        guard let baseAddress = buffer.baseAddress else {
            return 0
        }

        return proc_pidpath(getppid(), baseAddress, UInt32(buffer.count))
    }

    guard pathLength > 0 else {
        return false
    }

    return String(cString: parentPath).contains("/loginwindow.app/")
}

@_cdecl("chiri_macos_login_item_status")
public func chiriMacOSLoginItemStatus() -> Int32 {
    guard #available(macOS 13.0, *) else {
        return LoginItemStatus.unsupported.rawValue
    }

    return mainAppServiceStatus().rawValue
}

@_cdecl("chiri_macos_login_item_enable")
public func chiriMacOSLoginItemEnable(
    _ outputError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard #available(macOS 13.0, *) else {
        setError(outputError, "Launch at login requires macOS 13 or later.")
        return 0
    }

    let status = mainAppServiceStatus()
    if status == .enabled || status == .requiresApproval {
        return 1
    }

    do {
        try SMAppService.mainApp.register()
        return 1
    } catch {
        setError(outputError, error.localizedDescription)
        return 0
    }
}

@_cdecl("chiri_macos_login_item_disable")
public func chiriMacOSLoginItemDisable(
    _ outputError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard #available(macOS 13.0, *) else {
        setError(outputError, "Launch at login requires macOS 13 or later.")
        return 0
    }

    let status = mainAppServiceStatus()
    if status == .notRegistered || status == .notFound {
        return 1
    }

    do {
        try SMAppService.mainApp.unregister()
        return 1
    } catch {
        setError(outputError, error.localizedDescription)
        return 0
    }
}
